awx/main/models/events.py

```diff
@@ -605,19 +605,21 @@ def _update_host_summary_from_stats(self, hostnames):
     def _update_host_metrics(updated_hosts_list):
         from awx.main.models import HostMetric  # circular import
 
-        # bulk-create
         current_time = now()
-        HostMetric.objects.bulk_create(
-            [HostMetric(hostname=hostname, last_automation=current_time) for hostname in updated_hosts_list], ignore_conflicts=True, batch_size=100
-        )
-        # bulk-update
-        batch_start, batch_size = 0, 1000
-        while batch_start <= len(updated_hosts_list):
-            batched_host_list = updated_hosts_list[batch_start : (batch_start + batch_size)]
-            HostMetric.objects.filter(hostname__in=batched_host_list).update(
-                last_automation=current_time, automated_counter=models.F('automated_counter') + 1, deleted=False
-            )
-            batch_start += batch_size
+
+        # FUTURE:
+        #   - Hand-rolled implementation of itertools.batched(), introduced in Python 3.12.  Replace.
+        #   - Ability to do ORM upserts *may* have been introduced in Django 5.0.
+        #     See the entry about `create_defaults` in https://docs.djangoproject.com/en/5.0/releases/5.0/#models.
+        #     Hopefully this will be fully ready for batch use by 5.2 LTS.
+
+        args = [iter(updated_hosts_list)] * 500
+        for hosts in zip(*args):
```

Comment:
What if there are _fewer_ than 500 hosts? This "Hand-rolled" iterator is hard to read, but that's not a blocker. I tried to test the method in shell:

```python
>>> my_list = list(range(70))
>>> args = [iter(my_list)]*200
>>> for sublist in zip(*args):
...   print([f for f in sublist])
...   print([f for f in sublist])
... 
```

In this case, as `200 > 70` (imagine you have 70 hosts, and the batch size is 200), it fails to print anything. It seems that the main `for` loop doesn't process _any_ items.

Additionally, when I tested this with 70 list entries and a batch size of 20, it ended at 59, which is less than 70. I expected it to end at 69.

So both examples are suggesting to me that this method loses the partial batch at the end.

Reply:
@AlanCoding yeah, you're right.

Reply:
fix pushed